lib/charms/synapse/v1/matrix_auth.py

```diff
@@ -0,0 +1,530 @@
+# Licensed under the Apache2.0. See LICENSE file in charm source for details.
+
+"""Library to manage the plugin integrations with the Synapse charm.
+
+This library contains the Requires and Provides classes for handling the integration
+between an application and a charm providing the `matrix_plugin` integration.
+
+### Requirer Charm
+
+```python
+
+from charms.synapse.v0.matrix_auth import MatrixAuthRequires
+
+class MatrixAuthRequirerCharm(ops.CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.plugin_auth = MatrixAuthRequires(self)
+        self.framework.observe(self.matrix_auth.on.matrix_auth_request_processed, self._handler)
+        ...
+
+    def _handler(self, events: MatrixAuthRequestProcessed) -> None:
+        ...
+
+```
+
+As shown above, the library provides a custom event to handle the scenario in
+which a matrix authentication (homeserver and shared secret) has been added or updated.
+
+The MatrixAuthRequires provides an `update_relation_data` method to update the relation data by
+passing a `MatrixAuthRequirerData` data object, requesting a new authentication.
+
+### Provider Charm
+
+Following the previous example, this is an example of the provider charm.
+
+```python
+from charms.synapse.v0.matrix_auth import MatrixAuthProvides
+
+class MatrixAuthProviderCharm(ops.CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.plugin_auth = MatrixAuthProvides(self)
+        ...
+
+```
+The MatrixAuthProvides object wraps the list of relations into a `relations` property
+and provides an `update_relation_data` method to update the relation data by passing
+a `MatrixAuthRelationData` data object.
+
+```python
+class MatrixAuthProviderCharm(ops.CharmBase):
+    ...
+
+    def _on_config_changed(self, _) -> None:
+        for relation in self.model.relations[self.plugin_auth.relation_name]:
+            self.plugin_auth.update_relation_data(relation, self._get_matrix_auth_data())
+
+```
+"""
+
+# The unique Charmhub library identifier, never change it
+LIBID = "ff6788c89b204448b3b62ba6f93e2768"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 1
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 3
+
+# pylint: disable=wrong-import-position
+import json
+import logging
+from typing import Dict, List, Optional, Tuple, cast
+import secrets
+import base64
+from cryptography.fernet import Fernet
+
+import ops
+from pydantic import BaseModel, Field, SecretStr
+
+logger = logging.getLogger(__name__)
+
+#### Constants ####
+APP_REGISTRATION_LABEL = "app-registration"
+APP_REGISTRATION_CONTENT_LABEL = "app-registration-content"
+DEFAULT_RELATION_NAME = "matrix-auth"
+SHARED_SECRET_LABEL = "shared-secret"
+SHARED_SECRET_CONTENT_LABEL = "shared-secret-content"
+AES_KEY_SECRET_LABEL = "aes-key-secret"
+AES_KEY_SECRET_CONTENT_LABEL = "aes-key-content"
```

Comment:
We should probably rename all variables, names related to the encryption to avoid being "implementation" specific, and therefore not specifying "AES" but more "ENCRYPTION_KEY...."